### PR TITLE
[Review Only] public BehaviorBase, InputMap

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Reviewing prototype code for public BehaviorBase and InputMap, using a subset of controls:

ColorPicker
ComboBox
DatePicker
PasswordField
TextArea
TextField

### CODE REVIEW

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+ *    behavior methods, map key bindings to function tags, and add additional event handlers,
+ *    using func(), key(), and hand() methods correspondingly.
+ *    Important: no mapping should be made in the behavior constructor, only in install().
+ * <p>
+ * The base class adds a dispose() method (called from Skin.dispose()),
+ * which undoes the mapping done in install().
+ * <p>
+ * TODO rename BehaviorBase and remove IBehavior
+ */
+public abstract class BehaviorBase2<C extends Control> implements IBehavior {
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: `com.sun.javafx.scene.control.behavior.BehaviorBase<T>` class is a generic class where T =` N extends Node`. Here, we have T = `C extends Control`.
I think, having a `Behavior` specific for a `Control` is logically correct. 
Are we going to miss out on something which is in a `Behavior` class with type `Node`?

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    protected final InputMap2 getInputMap2() {
+        return control.getInputMap2();
+    }
+
+    /**
+     * Installs this behavior.
+     * This method must be called in Skin.install() to actually install all the default mappings.
+     * @param skin the skin
+     */
+    public void install(Skin<C> skin) {
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: In my opinion, `Behavior` class need not refer to `Skin` class at all. We can simply pass in the required `Control`.
A `Skin` is the owner of `Behavior` and can pass in the necessary `Control`

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextAreaBehavior.java [hunk 3]
```diff
@@ -144,6 +80,68 @@ public TextAreaBehavior(final TextArea c) {
         if (Utils.isTwoLevelFocus()) {
             tlFocus = new TwoLevelFocusBehavior(c); // needs to be last.
         }
+
+        // functions
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextAreaBehavior.java:85]: Earlier we used to have KeyMapping mapping a key (or combination) to a function directly.
Now, there is an indirection : Key-> FunctionTag and then FunctionTag -> Runnable function.

I am unable to envisage the need to have this indirection. Why are we trying to separate this?

It is one thing to provide flexibility, but I see this may lead to potential misunderstanding of the InputMap APIs and eventually more bugs!

File: modules/javafx.controls/src/main/java/javafx/scene/control/TextArea.java [hunk 4]
```diff
@@ -81,6 +82,51 @@
  * @since JavaFX 2.0
  */
 public class TextArea extends TextInputControl {
+    /** Denotes the function that moves the caret to the document end. */
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TextArea.java:85]: At a glance, this looks like we are polluting this `Control` class with `Behavior` related code.
This is another reason why I did not like the indirection of Key -> FunctionTag -> function.
Ideally, there should be no change to `Controls` classes while making `Behavior` classes public.

Imagine how weird the Javadoc for this class will look as we list all these FunctionTags.

**Review by aghaisas — COMMENTED** [review]
I have taken a quick look of proposed changes. I have a few questions.
A detailed review will be done later.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java (hunk 1 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: I think Control is the best candidate to have an InputMap.  There is currently no reason to have it in Node, since all the instantiations of the existing InputMap are done with Controls.  Same applies to (current) Behaviors.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java (hunk 2 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: You are right, strictly speaking.

The only reason to have a Skin argument here is to  implicitly suggest that:
a) this method must be invoked from Skin.install() and
b) both control and skin are both not null

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextAreaBehavior.java (hunk 3 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextAreaBehavior.java:85]: Thank you for a good question!
This is actually an important feature of the new design: being able to 
a) change a key binding while keeping the existing function (and back) and
b) change a function while keeping the existing key binding (and back).

This design is reminiscent of the Swing's ActionMap, except we introduce a `FunctionTag` and they use a simple String.

The use case is to support redefining the editor keys (in Eclipse, Settings -> General -> Keys), as well as implementing new behavior without subclassing the behavior class at all, for example the app/component developer can simply supply their own the copy or navigation functions without messing with the platform-specific  key bindings.  Want to skip over HTML tags when using arrow left / arrow right?  Redefine function mapped to TextInputControl.LEFT/RIGHT.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TextArea.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TextArea.java:85]: Interesting point!

One way to think about it is that each Control comes with some functionality by default.  We currently have copy() and paste() as public methods in, for instance, TextInputControl, and we should have public other methods that eventually go through the behavior.

In fact, part of the new design is making all of these methods public at the Control level, albeit going through indirection using `FunctionTag`s.  I think this idea makes the control more easily extendable.

I wish method pointers were more static in java, we could use them, but TextInputControl::copy creates a new lambda instance each time, thus we need to declare a method identifier explicitly.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+        en.behaviorValue = handlers;
+    }
+    
+    /**
+     * Adds a user-specified function under the given function tag.
+     * This function will override any function set by the behavior.
+     * @param tag the function tag
+     * @param function the function
+     */
+    public void func(FunctionTag tag, Runnable function) {
```
Comment (karthikpandelu) [inline modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java]: Do you think having the function name like `setFunction` would be better than `func`?

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+    }
+    
+    /**
+     * Link a key binding to the specified function tag.
+     * This method will override a mapping set by the behavior.
+     *
+     * @param k the key binding
+     * @param tag the function tag
+     */
+    public void key(KeyBinding2 k, FunctionTag tag) {
```
Comment (karthikpandelu) [inline modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java]: Similar to previous comment, can we change this to easily understandable name since it is a public method? Like `setKey` or `mapKeyBinding`?

**Review by karthikpandelu — COMMENTED** [review]
Some of the questions I had has been already answered in the previous. Added couple of comments.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java (hunk 5 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java]: good question!  I've tried to make it shorter in order to fit the registration code in the behaviors to a single line
(see for example TextInputControlBehavior 114-208).
May be regFunc() and regKey() ?

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java (hunk 6 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java]: also, 
map() -> addHandler()
mapTail() -> addHandlerTail()
or something?

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java (hunk 5 above)
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java]: > good question! I've tried to make it shorter in order to fit the registration code in the behaviors to a single line (see for example TextInputControlBehavior 114-208). 

Ok, understood. Looks like in few cases line length might become more than what we usually keep according to coding guidelines.

>May be regFunc() and regKey() ?

This also sounds good.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java (hunk 6 above)
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java]: Sure. I believe it will make it easy to understand the API functionality.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextAreaBehavior.java (hunk 3 above)
Comment (aghaisas) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextAreaBehavior.java:85]: OK. Now, I see what the new design is trying to address.

I do not fully agree with your statement "as well as implementing new behavior without subclassing the behavior class at all". If a user intends to change the behavior of a control, he/she should do it by explicitly sub-classing (soon to be public) Behavior APIs.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TextArea.java (hunk 4 above)
Comment (aghaisas) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TextArea.java:85]: If one reads these function tag names, then it is obvious that they refer to the Behavior.
Hence, according to me, the correct place for them is in Behavior classes.
The fact that they are defined here and initialized in corresponding Behavior classes corroborates this.

If a user wants to customize the Behavior, he/she should do it by providing a custom Behavior class explicitly.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextAreaSkin.java [hunk 7]
```diff
@@ -828,16 +820,29 @@ public Bounds getCaretBounds() {
...
-    @Override public void dispose() {
-        if (getSkinnable() == null) return;
-        getSkinnable().removeEventFilter(ScrollEvent.ANY, scrollEventFilter);
-        getChildren().remove(scrollPane);
-        super.dispose();
-
-        if (behavior != null) {
-            behavior.dispose();
+        // install default input map for the text area control
+        behavior = new TextAreaBehavior();
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextAreaSkin.java]: I think, we can create the `behavior` object in the `Skin` constructor and invoke `Behavior.install()` from `Skin.install()` method.
It is possible to erroneously invoke this `install` method twice and we would end up with a leak.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+     * Maps a key binding to the specified function tag.
+     * This method will not override a user mapping added by {@link #regKey(KeyBinding2,FunctionTag)}.
+     *
+     * @param code the key code to construct a {@link KeyBinding2}
+     * @param tag the function tag
+     */
+    protected void regKey(KeyCode code, FunctionTag tag) {
+        getInputMap2().regKey(this, code, tag);
+    }
+
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: Do you think whether it makes sense to add a variant - `regKeyAndFunc(KeyCode/KeyBinding, FunctionTag, Runnable)` for ease of use while defining default mappings?

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: The earlier versions had such a method.
We could add it; I just thought that
a) the resulting code might easily break 120 character limit and
b) it might be better to have a separate block that defines the functions, since it cannot have any platform conditionals.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextAreaSkin.java (hunk 7 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextAreaSkin.java]: this makes sense, thanks!

**Comment by andy-goryachev-oracle** [comment]
> Hence, according to me, the correct place for them is in Behavior classes.

No, not really.  The function tags are exposed at the Control level, to allow for redefining of either key mapping or the function itself.

For example, the user might decide to redefine a key binding (or add another) mapped to some function.   This will be done at the control level.  Or another case, when the user needs to change the function itself without modifying the key bindings - it's also done at the control level.

Note that this PR does not completely implements the new paradigm (yet).

**Comment by andy-goryachev-oracle** [comment]
> If a user intends to change the behavior of a control, he/she should do it by explicitly sub-classing (soon to be public) Behavior APIs.

It depends - what I am saying some degree of customization might be possible without subclassing, depending on the requirements.  For example:
1. changing a key binding does not require subclassing
2. replacing the default function with something that can be implemented using public APIs does not require subclassing.

Consider, for example, a case when the user wants to change navigation (left arrow / right arrow) to skip over some tokens.  The user might implement this new functionality using getText() and select() methods without subclassing, and simply registering the new function.  Anything more involved, anything that requires deeper changes - yes, the user will need to subclass.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java (hunk 8 above)
Comment (aghaisas) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: (a) should not be a concern as we can break the 3 arguments over 3 lines if needed.
(b) makes sense
We can add it if it is going to make our life easier while implementing all the default Behaviors.

**Comment by aghaisas** [comment]
> It depends - what I am saying some degree of customization might be possible without subclassing, depending on the requirements. For example:
> 
>     1. changing a key binding does not require subclassing
> 
>     2. replacing the default function with something that can be implemented using public APIs does not require subclassing.
> 
> 
> Consider, for example, a case when the user wants to change navigation (left arrow / right arrow) to skip over some tokens. The user might implement this new functionality using getText() and select() methods without subclassing, and simply registering the new function. Anything more involved, anything that requires deeper changes - yes, the user will need to subclass.

I think we are mixing ease of use vs custom behavior.
If we take the example that you have mentioned about "changing navigation with left arrow/right arrow to skip over some tokens" - this indicates the user would like to change the default behavior of those keys and in my opinion it should be explicitly called out using custom Behavior class.
This is the reason all those public function tags should be part of corresponding Behavior class of a Control and **not** part of `Control` class.

**Comment by aghaisas** [comment]
My understanding is - A default `Skin` object instantiates and contains a default `Behavior` object.
How should a Custom Behavior class be written by a developer? Do they have to provide a corresponding Custom Skin as well?

**Comment by andy-goryachev-oracle** [comment]
> How should a Custom Behavior class be written by a developer? Do they have to provide a corresponding Custom Skin as well?

A very good question!

A typical concrete Behavior might contain two things: the methods (==functions) that embody the default implementation, and the mapping between input events and said functions.  This is not an iron-clad rule, as the design also allows for adding user-defined mappings (to method located elsewhere).

In order to change the behavior, the mappings must be changed.  The next question is how involved the change is.  If the change is simple and the new function can be implemented by a piece of code that requires nothing but public APIs available via control or skin or default implementation, then a custom skin may simple map the new behavior, overriding the one provided by the stock behavior.

If, however, the change is more involved and requires deep intrusion into bowels of the skin and/or behavior, then we must make both skin and behavior sufficiently customizable and public, **which is not how we've done things in the past**.  An example would be something like adding multiple selection to a text component - there is no way this can be done without essentially providing a new skin, a new behavior class, and changing some of the selection APIs.  Would that be a complete re-write?  Quite possibly.

To summarize, this proposal allows the user to customize the behavior to a certain degree without the need to subclass its behavior, and even without subclassing the skin, if the new functionality can be implemented with existing public APIs in the control or skin or the default behavior.

File: modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java (hunk 8 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java]: added methods, thanks!

**Comment by andy-goryachev-oracle** [comment]
closed in favor of https://github.com/openjdk/jfx/pull/1254
